### ISSUE TITLE
Django + heroku for backend

### ISSUE THREAD

**Issue by hdoupe** [DataKind-DC/homelessness-service-navigator#5]
The consensus after the Code for DC meet up on 12/13 was that we should use Django and Heroku for the backend.  However, the frontend developers who have already done some good work were not there.  So, I am particularly interested in their thoughts on using this stack.  Further, it's helpful to have these types of discussions catalogued for future reference. 

We primarily chose to use this stack since this is what the more backend oriented contributors have experience in.  Also, it's easy to setup and maintain.  Here's a [link](https://devcenter.heroku.com/articles/deploying-python) to the heroku docs for how to set this up.  I recommend that those who have not used Django before work through some of the [tutorial](https://docs.djangoproject.com/en/2.0/intro/tutorial01/).  I think you'll find that it's ridiculously easy to get a Django app up and running.

**Comment by geoffreyyip** [comment]
From Slack:

Geoffrey Yip 4:24 PM
Curious, why django?

Django's very magic-heavy, and designed for medium-sized publishers (Atlantic Media's a notable ). I fear it'd be overkill for our website.

I'd be comfortable if Code for DC was known for its Django community, or if we had a Django programmer committed to our project.

williamratcliff 5:50 PM
Hank added an issue on Github to explain the basic reasoning.   He's made some apps for it and so have I (though using earlier versions, so I would need to see how things have changed).  I would say that it's no more magic heavy than say Rails.   We have at least a few people that either develop with it or are familiar with it.  Our first thought was that for the backend developers, python was a common language.   It's also likely to be used for the routing portions of the app that would follow.

williamratcliff 5:57 PM
Now, for the initial front end we probably could have used something light weight, like bottle (or flask).    But applications have an ugly habit of growing.   Chatting with Theresa, at some point she suggested it would be nice for the users to establish accounts and save their information (for example, which facilities they plan to visit).   Because they will be using library and other public computers, she suggested that we couldn't just use local storage because they will want to maintain privacy.   So, then we're dealing with a user system.  We could end up building all of this, but it seems like it's just easier to use Django with batteries included (the learning curve is pretty quick for basic things and I've had pretty good luck bringing high school students and college students up to speed quickly).

5:59
Where I see more of a difficulty is that at least earlier versions of django really felt geared more towards web 1.0.  There is django rest framework, but a year or two ago, we were playing with integrating django and react for an application and I have to admit that we kludged a lot together.

6:02
We considered Node, but the thoughts of doing anything with data science and routing later made us favor python.

Geoffrey Yip 8:49 PM
Got it, I'm copying our response over to the Github issue to have it recorded.

**Comment by geoffreyyip** [comment]
I'm unfamiliar with back-end, but am okay with this decision since @williamratcliff said we have a few Django developers.

Personally found Django to have a heavy learning curve. I "helped" out a Move On engineer with their open-source project with Django way back. And by help, I mean, he brought me coffee and lunch while he bestowed programming knowledge on me. If he wasn't doing that, I don't think I would have stuck with it.

**Comment by hdoupe** [comment]
@geoffreyyip Thanks for contributing to this discussion.  You raise some good points, and I will try to address them below.

@geoffreyyip said
> Django's very magic-heavy, and designed for medium-sized publishers (Atlantic Media's a notable ). I fear it'd be overkill for our website.

My interpretation of Django being "magic-heavy" is that Django takes care a lot of website functionality so that we don't have to.  My only experience with web development was gained after taking over my team's Django app a few months ago.  I have learned a lot about web development in these months, but I'm definitely not ready to write a scalable app using a lower level framework such as Flask.  If someone else in the project has this type of experience and wants to start with Flask or some other framework, then I would be interested in helping out and learning from them.  Otherwise, this doesn't disqualify Django for me.

Django is used by publishers like Atlantic Media and the Washington Post.  However, I think that this speaks more to Django's flexibility.  You can set up a Django app that begins small and grows into something larger if it needs to.  If we started with a lightweight framework, then we would have to revisit this decision down the road when we need to add login capabilities or routing functionality for example.

There is a bit of a "learning curve," but I think it's worth it.  There's a ton of documentation, and in my opinion, enough of us have sufficient Django experience to start this project and guide those with less exposure to Django.

**Comment by electricflannel** [comment]
Hey there!  I understand the need for a backend for user authentication, but why would we choose Django over a basic node app? 

All we need is user authentication, which Node can handle. As for scalibility, node performs in that realm as well.  

Something else to consider is that we will use lose the accessibility that JavaScript provides for both beginners and experienced folks because we are using the same language throughout the whole stack.

I do understand that Python might be good for data science. However, I'm not sure that we would leverage Python's capabilities for something we aren't working with yet. I know we talked about maybe doing some mapping and routing which can be done with Python, but that's no trivial task. There are plenty of APIs we could use for mapping that will probably work better than a bespoke solution.

These are just some things I think we should consider before switching out the original proposed plan.

**Comment by williamratcliff** [comment]
Excellent question!   Life everything, it's a question of trade-offs.    Before the suggestion of a user, I might have been tempted to run serverless.   However, with that requirement (which I don't see a good way around given that people are likely to be accessing this application from public computers as mentioned above), I think we're forced into having a server.   On the one hand, we have front-end developers who are expert at Javascript.  On the other hand, we have data scientists (this is a joint project between DataKind and Code for DC) who are far more comfortable with python.   It was also the choice of the backend developers who were present at the last Code for DC meetup.  As I look past the initial requirements of the service navigator, we should think about what our next steps are.  

1)  Routing--this will happen.   Is anyone aware of an API?  I am not.   However, I'm quite confident that Jess can come up with a first order solution (and he's started thinking about this).    

2)  Theresa brought up  the idea of visualizing data (this is not in place yet) on usage overlaid with the location of service areas.   This sounds like a question of using mapbox, but this is something for down the pike (along with predicting spikes in service usage, etc. which DataKind has worked on in the UK based on their data)

We could go the route of microservices, with primary server functionality in Node and data science portions in python, but this seems even more complex.

The difficulty that I see is not that python being different from javascript forming a barrier to people who want to join the project.   The main challenge I see is philosophical.   As I mentioned Django (at least 1.x is designed more around a Web 1.0 model, I will have to look at 2.0 which was recently released).    The front end for this is initially fairly simple, but we have to decide on things such as whether we handle routing on the client side or on the server side (My leaning would be to handle it on the client side).  Do we want to develop a single page application (my leaning is yes)?    Where do we handle filtering logic (since this is slowly changing, I would also tend to go with the client)?

To summarize, the decision was made due to a combination of future proofing and expertise.    I think in version 0, our goal is to simply deploy to Heroku and two of us are experienced in doing this with Django apps.

**Comment by geoffreyyip** [comment]
I disagree with this decision, and see the additional requirements as scope
creep. However, our potential user wants user accounts and you guys seem to
be sold on Django.

I'm neutral between Flask and Django, but I'm skeptical of the need for a
back-end, given that our underlying data (a table of homeless services) has
infrequent updates. The table is small.enough to be hosted in the repo, or
in a low-config Heroku add-on.

I do like Django though and would like to learn more of it. However, it
lies outside expertise and I'll need someone else to set up the boilerplate
and issue me tickets before I can contribute.

**Comment by williamratcliff** [comment]
I wouldn't say scope creep so much as talking with our users.  I think in
the short term, our choice of server technology won't matter if we agree to
do as much client side for now as possible.

**Comment by williamratcliff** [comment]
One thing I think we all can do is ask ourselves what would convince us
that we were wrong

**Comment by electricflannel** [comment]
Regardless of backend choice, I'm happy to help out where I can

**Comment by hdoupe** [comment]
Thanks for the discussion. It looks like we are going to start with Django 1.11.

**Comment by hdoupe** [comment]
Closed #5
